### ISSUE TITLE
skimr v2 available for testing: Important API changes

### ISSUE THREAD

**Issue by michaelquinn32** [ropensci/skimr#341]
Hi everyone!

@elinw and I have received a lot of great feedback on `skimr` over the last year, and we've done our best to incorporate all of it into the next version of `skimr`: v2. This is a major overhaul of both the `skimr` API and internals. You can see the complete design here:

https://docs.google.com/document/d/18lBStDZzd1rJq08O-4Sw2qHhuHEZ79QX4sBkeyzWNFY/edit?usp=sharing

The current version of the code is here:

https://github.com/ropenscilabs/skimr/tree/v2

Some highlights include:

*  `skim_with()` now returns a function instead of modifying a global state. This is safer and lets you have multiple `skim()` functions in a session.
* A new type of object `sfl()` (a `skimr` function list) is used within `skim_with()`. It can automatically add names to functions and allows for pronouns like in `dplyr::funs()`.
* We have a new API for providing your own default summary functions: `get_skimmers()`. You can
export a method for this generic in a separate package to extend `skimr`.
* The output of `skim` now better matches user expectations: `skim(iris) %>% dplyr::filter(sd > 1)` now works.
* We have better integration with `knitr`, through the `knit_print()` API. `skimr` output should render nicely without extra work.
* `partition()` takes a `skim_df` and produces a list of simplified subtables, by type. `yank()` lets you get a single subtable. This should make customizing output a little simpler.

To get this, we've needed some breaking changes:

* We can't support skimming individual vectors. Data needs to be in a `data.frame()` instead.
* Our versions of `kable` and `pander` have been deprecated.
* Other helper functions have been deprecated, including those for reshaping `skim()` output and investigating global options.

To top it off, `skimr` has become much simpler and is much better tied to the current state of the `tidyverse`. Our hope is that is will be much easier to maintain going forward.

I imagine that this preview version might still be a little rough around the edges, but please give it a shot so we can catch bugs before the release. You can install it with `devtools`:

```
devtools::install_github("ropenscilabs/skimr", ref = "v2")
```

Last, and especially not least, a big thank you goes to @GShotwell, @akraemer007, @puterleat, @tonyfischetti, @Nowosad, @rgayler, @jrosen48, @randomgambit, @elben10, @koliii, @AndreaPi, @rubenarslan, @jenniferthompson and @jimhester for all of the great feedback over the last year. We couldn't have done this without your input.

**Comment by koliii** [comment]
Congratulations! Michael :)

**Comment by elinw** [comment]
This is exciting.  So what I've done is to create a **release_v1** branch that can serve as the de facto develop branch for the currently released version.  Then changes based on the road map can be merged into develop but not prevent us from doing bug fixes and enhancements for v1 and merging them into master.  That does mean we  have to  make sure that we get things, especially bug fixes and history, into both branches.

**Comment by AndreaPi** [comment]
Hi, I still fumble every now and then when modifying the list of skimmers. What would be the best place to ask support for this kind of things? Here https://discuss.ropensci.org/ or Stack Overflow?

**Comment by michaelquinn32** [comment]
I think here is the best place. It's probably the one we check most often.

We'd be happy to help however possible.

**Comment by GegznaV** [comment]
In **version2 **, consider not to print the results as separate tibbles, as the result gets not as compact as in **version 1** and it is polluted with unnecessary information (I highlighted the unnecessary lines):

![image](https://user-images.githubusercontent.com/12725868/44203565-123d4780-a158-11e8-8100-27560b471a8f.png)

**Comment by michaelquinn32** [comment]
Thanks!

I'll dig deeper into how tibbles are formatted to try and resolve some of this. I'm not fully clear on all of this yet, but we might still be blocked by
https://github.com/tidyverse/tibble/issues/364

I won't drop support for tibbles/ pillar. The current version of skimr needs to do a bunch of ugly stuff to (barely) support unicode on different platforms. Pillar provides this support for us, so we can drop all of that code.
https://github.com/tidyverse/tibble/issues/316

**Comment by jimhester** [comment]
@michaelquinn32 You could filter these by calling `format()` first then manually removing the lines, although I agree having explicit support for this in tibble would be much better. e.g.

``` r
mtcarst <- tibble::as_tibble(mtcars)

out <- format(mtcarst)
cat(out[c(-1, -3)], sep = "\n")
#>      mpg   cyl  disp    hp  drat    wt  qsec    vs    am  gear  carb
#>  1  21       6  160    110  3.9   2.62  16.5     0     1     4     4
#>  2  21       6  160    110  3.9   2.88  17.0     0     1     4     4
#>  3  22.8     4  108     93  3.85  2.32  18.6     1     1     4     1
#>  4  21.4     6  258    110  3.08  3.22  19.4     1     0     3     1
#>  5  18.7     8  360    175  3.15  3.44  17.0     0     0     3     2
#>  6  18.1     6  225    105  2.76  3.46  20.2     1     0     3     1
#>  7  14.3     8  360    245  3.21  3.57  15.8     0     0     3     4
#>  8  24.4     4  147.    62  3.69  3.19  20       1     0     4     2
#>  9  22.8     4  141.    95  3.92  3.15  22.9     1     0     4     2
#> 10  19.2     6  168.   123  3.92  3.44  18.3     1     0     4     4
#> # ... with 22 more rows
```

Created on 2018-08-20 by the [reprex package](http://reprex.tidyverse.org) (v0.2.0).

**Comment by elinw** [comment]
This is to me (along with losing rownames) the biggest disadvantage of using tibbles so I'm really happy to have a relatively easy solution from @jimhester.    I don't think we're actually using pillar at all, however.  We did initially but since it was not a stable API and no longer exported  the pillar team contributed the code to skimr.

**Comment by elinw** [comment]
@AndreaPi Actually I do periodically look for skimr questions on Stack Overflow, there are actually a number of them.  I wonder if we could get a tag.

**Comment by jacob-long** [comment]
I would suggest adding some documentation (unless I've missed it) for changing the number of digits past the decimal are printed in v2. I'm not certain how this is changed in interactive use, but it looks like from digging several internal functions deep that the knitted output is affected by a "skimr_digits" option.

**Comment by elinw** [comment]
I'm working in a branch on updating the tests and a few other things to work with the newest version of dplyr (0.8.0).   However this does lead me to a few thoughts, keeping in mind that of course the outputs from dplyr change because of the changed skim_df object but also that for us a key change in dplyr is that it no longer strips the skim_df class from the results.  This meant that for v1 I implemented skimr stripping of the class so that things work as previously (and that will be released prior to dpylr 0.8.0 if all goes well).

1. It might make sense in v2 to retain the skim_df class after processing with dplyr.  Mainly (I think) this handles the print, which would be a skim print rather than a data frame print.  That is too radical a change for v1 (where people may be counting on the data frame print in their markdown files, as one example). I would rather do that explicitly as a change from the v1 API rather than as part of the current PR.  But I think it could be useful to know that this is a product of skimming and we can have some functions that could be used on a post dplyr data frame.

2. This does really highlight the fact that skim_df is totally different in v2.  What do we do about people's use cases that really count on the skim_df structure? I wonder if it is really right that they have the same names when they are so different.  I  don't really want skim_df2 because it's ugly.  Maybe we should add a version attribute to v2 (and maybe to v1 -- I don't see how it hurts anything, but I could be missing something). Then at least it's possible to check.  We'll have to keep in mind that there are 4 combinations of skimr version and dplyr version (before 0.8.0 and 0.8.0 or later) that we'll need to cope with. 

3. This makes me wonder if there shouldn't be a version parameter on some functions.  For example, we could make the presence of the skim_df class on the processed data dependent on the version attribute.

4. Related to that we definitely need a way for v1 users to get a long object from v2.   
```
x <- skim(iris)
y <- tidyr::gather(x, key = "name", value = "formatted", -"type", -"variable") %>% 
            dplyr::arrange(desc(variable))  %>% rename( stat = name)
```   
gets us almost there and is so easy now because of the v2 object being so easy to work with.  

@michaelquinn32

**Comment by michaelquinn32** [comment]
Here's the updated V2 output after https://github.com/ropensci/skimr/commit/627214fdea54bea628750f7b960ef49c4f49640a

```
> options(width = 85)
> library(skimr)
> skim(iris)
Skim summary statistics
 n obs: 150
 n variables: 5
── Variable type: factor ────────────────────────────────────────────────────────────
  variable missing complete     n ordered n_unique top_counts
1 Species        0      150   150 FALSE          3 set: 50, ver: 50, vir: 50, NA: 0
── Variable type: numeric ───────────────────────────────────────────────────────────
  variable     missing complete     n  mean    sd    p0   p25   p50   p75  p100 hist
1 Sepal.Length       0      150   150  5.84 0.828   4.3   5.1  5.8    6.4   7.9 ▆▇▇▅▂
2 Sepal.Width        0      150   150  3.06 0.436   2     2.8  3      3.3   4.4 ▁▆▇▂▁
3 Petal.Length       0      150   150  3.76 1.77    1     1.6  4.35   5.1   6.9 ▇▁▆▇▂
4 Petal.Width        0      150   150  1.20 0.762   0.1   0.3  1.3    1.8   2.5 ▇▁▇▅▃
```

**Comment by elinw** [comment]
One important change we made yesterday is to require a minimum dplyr version of 0.8.0 for v2.  This means some users will probably stay on version 1 longer than they and we would like, but it reduces the complexity a lot.

**Comment by michaelquinn32** [comment]
We're now in the process of moving v2 towards release. Thanks everyone for all of the feedback along the way! (#383)
